File: 03-java-socket/src/main/java/org/protelis/demo/HelloProtelis.java
package org.protelis.demo;

import com.uchuhimo.konf.BaseConfig;
import com.uchuhimo.konf.Config;
import com.uchuhimo.konf.source.DefaultLoaders;
import com.uchuhimo.konf.source.DefaultTomlLoaderKt;
import com.uchuhimo.konf.source.Loader;
import org.protelis.demo.data.ProtelisNode;
import org.protelis.lang.ProtelisLoader;
import org.protelis.vm.ProtelisProgram;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Example usage of the implemented classes.
 */
public final class HelloProtelis {

    private HelloProtelis() { }

    /**
     * Main method.
     * @param args unused
     * @throws IOException in case of issues with local networking
     */
    public static void main(final String[] args) throws IOException {
        Config config = new BaseConfig();
        config.addSpec(ProtelisConfigSpec.SPEC);
        final DefaultLoaders defaultLoaders = config.from();
        final Loader toml = DefaultTomlLoaderKt.getToml(defaultLoaders);
        config = toml.resource("config.toml", false);
        final String protelisModuleName = config.get(ProtelisConfigSpec.protelisModuleName);
        final int iterations = config.get(ProtelisConfigSpec.iterations);
        final List<ProtelisNode> nodes = config.get(ProtelisConfigSpec.nodes);
        final List<Device> devices = new ArrayList<>();
        for (final ProtelisNode n: nodes) {
            final SocketNetworkManager netmgr = new SocketNetworkManager(
                new IntDeviceUID(n.getId()),
                n.getHostandport().getPort(),
                n.getNeighbors()
            );
            netmgr.listen();
            final ProtelisProgram program = ProtelisLoader.parse(protelisModuleName);
            final Device node = new Device(program, n.getId(), netmgr, new ConsoleSpeaker());
            if (n.isLeader()) {
                node.getDeviceCapabilities().getExecutionEnvironment().put("leader", true);
            }
            devices.add(node);
        }
        // Run some cycles
        for (int i = 0; i < iterations; i++) {
            devices.forEach(Device::runCycle);
        }
        // Stop the server socket
        devices.forEach(d -> ((SocketNetworkManager) d.getNetworkManager()).stop());
    }
}

File: 06-kotlin-mqtt/src/main/kotlin/org/protelis/demo/HelloProtelis.kt
package org.protelis.demo

import com.uchuhimo.konf.Config
import com.uchuhimo.konf.source.toml
import org.protelis.lang.ProtelisLoader

/**
 * Demonstration of the [MqttNetworkManager] usage.
 */
fun main() {
    var devices: List<Device> = emptyList()
    val config =
        Config { addSpec(ProtelisConfigSpec) }
            .from.toml
            .resource("config.toml")
    val protelisModuleName = config[ProtelisConfigSpec.protelisModuleName]
    val iterations = config[ProtelisConfigSpec.iterations]
    val nodes = config[ProtelisConfigSpec.nodes]
    // Initialize some nodes.
    nodes.forEach {
        val mqttNetworkManager =
            MqttNetworkManager(IntDeviceUID(it.id), neighbors = it.neighbors)
                .apply { listen(it.listen) }
        val program = ProtelisLoader.parse(protelisModuleName)
        val node = Device(program, it.id, mqttNetworkManager, ConsoleSpeaker())
        if (it.leader) {
            node.deviceCapabilities.executionEnvironment.put("leader", true)
        }
        devices += node
    }
    // Let the nodes make some iterations.
    repeat(iterations) { _ ->
        devices.forEach { it.runCycle() }
    }
    // Close the thread listening.
    devices.forEach { (it.networkManager as MqttNetworkManager).stop() }
}
